relay-redis/src/real.rs

```diff
@@ -382,3 +396,154 @@ pub struct Stats {
     /// The number of idle connections.
     pub idle_connections: u32,
 }
+
+/// [`bb8::ManageConnection`] for a async redis cluster.
+pub struct RedisClusterConnectionManager {
+    client: ClusterClient,
+}
+
+impl RedisClusterConnectionManager {
+    /// Creates a new [`bb8::ManageConnection`] with the provided [`ClusterClient`]
+    pub fn new(client: ClusterClient) -> Self {
+        Self { client }
+    }
+}
+
+impl bb8::ManageConnection for RedisClusterConnectionManager {
+    type Connection = ClusterConnection;
+    type Error = redis::RedisError;
+
+    fn connect<'life0, 'async_trait>(
+        &'life0 self,
+    ) -> Pin<Box<dyn Future<Output = Result<Self::Connection, Self::Error>> + Send + 'async_trait>>
+    where
+        'life0: 'async_trait,
+    {
+        let client = self.client.clone();
+        Box::pin(async move { client.get_async_connection().await })
+    }
+
+    fn is_valid<'life0, 'life1, 'async_trait>(
+        &'life0 self,
+        conn: &'life1 mut Self::Connection,
+    ) -> Pin<Box<dyn Future<Output = Result<(), Self::Error>> + Send + 'async_trait>>
+    where
+        'life0: 'async_trait,
+        'life1: 'async_trait,
+    {
+        Box::pin(async move { redis::cmd("PING").query_async(conn).await })
+    }
+
+    fn has_broken(&self, _: &mut Self::Connection) -> bool {
+        false
+    }
+}
+
+/// An abstraction over Single and Cluster redis connections.
+#[derive(Debug, Clone)]
+pub enum AsyncRedisPool {
+    /// Represents a connection pool to a clustered redis instance.
+    Cluster(bb8::Pool<RedisClusterConnectionManager>, RedisConfigOptions),
+    /// Represents a connection pool to a single redis server.
+    Single(
+        bb8::Pool<bb8_redis::RedisConnectionManager>,
+        RedisConfigOptions,
+    ),
+}
+
+impl AsyncRedisPool {
+    /// Takes a command and executes it on redis.
+    pub async fn query_async<T: FromRedisValue>(&self, cmd: Cmd) -> Result<T, RedisError> {
+        match self {
+            Self::Cluster(pool, options) => {
+                let mut conn = pool.get().await.map_err(RedisError::AsyncPool)?;
+                timeout(
+                    Duration::from_secs(options.read_timeout),
+                    cmd.query_async(&mut *conn),
+                )
```

Comment:
Since async redis doesn't seem to have `set_read_timeout`/`set_write_timeout` anymore. From what I understand both write and read timeouts are applied for every command while sending the command to redis and waiting for the reply respectively.
If that's true I think it would be better to use `timeout(read_timeout + write_timeout, ...)` to allow for the same total waiting as before

Reply:
Isn't this the same as `set_response_timeout` on the multiplexed connection or better, set connection and response timeouts on `AsyncConnectionConfig` and with `get_multiplexed_async_connection_with_config` this should do the right thing? I think we should just rely on that.

> If that's true I think it would be better to use timeout(read_timeout + write_timeout, ...) to allow for the same total waiting as before.

This still has different semantics. Socket read/write is independent from total duration.

Reply:
> The original thought still remains unclear for me, does it make sense to set the timeout to read + write?

I think just limiting it to `read` is fine. Let's just document the behaviour and change the config once all clients are async.